File: app/src/main/java/com/machiav3lli/backup/handler/action/RestoreSpecialAction.java
/*
 * OAndBackupX: open-source apps backup and restore app.
 * Copyright (C) 2020  Antonios Hazim
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.machiav3lli.backup.handler.action;

import android.content.Context;
import android.util.Log;

import com.machiav3lli.backup.Constants;
import com.machiav3lli.backup.handler.Crypto;
import com.machiav3lli.backup.handler.ShellHandler;
import com.machiav3lli.backup.items.AppInfo;
import com.machiav3lli.backup.utils.PrefUtils;

import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.stream.Collectors;

public class RestoreSpecialAction extends RestoreAppAction {
    private static final String TAG = Constants.classTag(".RestoreSpecialAction");

    public RestoreSpecialAction(Context context, ShellHandler shell) {
        super(context, shell);
    }

    private static boolean areBasefilesSubsetOf(File[] set, File[] subsetList) {
        Collection<String> baseCollection = Arrays.stream(set).map(File::getName).collect(Collectors.toCollection(HashSet::new));
        Collection<String> subsetCollection = Arrays.stream(subsetList).map(File::getName).collect(Collectors.toCollection(HashSet::new));
        return baseCollection.containsAll(subsetCollection);
    }

    @Override
    protected void restoreAllData(AppInfo app) throws Crypto.CryptoSetupException, RestoreFailedException {
        this.restoreData(app);
    }

    @Override
    public void restoreData(AppInfo app) throws RestoreFailedException, Crypto.CryptoSetupException {
        Log.i(TAG, String.format("%s: Restore special data", app));
        File backupDirectory = this.getDataBackupFolder(app);
        boolean isEncrypted = PrefUtils.isEncryptionEnabled(this.getContext());
        File archiveFile = this.getBackupArchive(app, BaseAppAction.BACKUP_DIR_DATA, isEncrypted);
        try {
            if (!archiveFile.exists()) {
                Log.i(TAG,
                        String.format("%s: %s archive does not exist: %s", app, BaseAppAction.BACKUP_DIR_DATA, archiveFile));
                return;
            }
            // uncompress the archive to the app's base backup folder
            this.uncompress(this.getBackupArchive(app, BaseAppAction.BACKUP_DIR_DATA, isEncrypted), this.getAppBackupFolder(app));

            // check if all expected files are there
            File[] filesInBackup = backupDirectory.listFiles();
            File[] expectedFiles = Arrays.stream(app.getFilesList()).map(File::new).toArray(File[]::new);
            if (filesInBackup != null && (filesInBackup.length != expectedFiles.length || !RestoreSpecialAction.areBasefilesSubsetOf(expectedFiles, filesInBackup))) {
                String errorMessage = String.format(
                        "%s: Backup is missing files. Found %s; needed: %s",
                        app, Arrays.toString(filesInBackup), Arrays.toString(expectedFiles));
                Log.e(TAG, errorMessage);
                throw new RestoreFailedException(errorMessage, null);
            }

            List<String> commands = new ArrayList<>(expectedFiles.length);
            for (File restoreFile : expectedFiles) {
                commands.add(this.prependUtilbox(String.format(
                        "cp -RLp \"%s\" \"%s\"",
                        new File(this.getDataBackupFolder(app), restoreFile.getName()),
                        restoreFile)));
            }
            String command = String.join(" && ", commands);
            ShellHandler.runAsRoot(command);

        } catch (ShellHandler.ShellCommandFailedException e) {
            String error = BaseAppAction.extractErrorMessage(e.getShellResult());
            Log.e(TAG, String.format(
                    "%s: Restore %s failed. System might be inconsistent: %s", app, BaseAppAction.BACKUP_DIR_DATA, error));
            throw new RestoreFailedException(error, e);
        } catch (IOException e) {
            Log.e(TAG, String.format(
                    "%s: Restore %s failed with IOException. System might be inconsistent: %s", app, BaseAppAction.BACKUP_DIR_DATA, e));
            throw new RestoreFailedException("IOException", e);
        } finally {
            boolean backupDeleted = FileUtils.deleteQuietly(backupDirectory);
            Log.d(TAG, String.format(
                    "%s: Uncompressed %s was deleted: %s", app, BaseAppAction.BACKUP_DIR_DATA, backupDeleted));
        }
    }

    @Override
    public void restorePackage(AppInfo app) {
        // stub
    }

    @Override
    public void restoreDeviceProtectedData(AppInfo app) {
        // stub
    }

    @Override
    public void restoreExternalData(AppInfo app) {
        // stub
    }

    @Override
    public void restoreObbData(AppInfo app) {
        // stub
    }
}

File: app/src/main/java/com/machiav3lli/backup/actions/RestoreSpecialAction.kt
/*
 * OAndBackupX: open-source apps backup and restore app.
 * Copyright (C) 2020  Antonios Hazim
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.machiav3lli.backup.actions

import android.content.Context
import com.machiav3lli.backup.dbs.entity.Backup
import com.machiav3lli.backup.dbs.entity.SpecialInfo
import com.machiav3lli.backup.handler.ShellHandler
import com.machiav3lli.backup.handler.ShellHandler.Companion.quote
import com.machiav3lli.backup.handler.ShellHandler.Companion.runAsRoot
import com.machiav3lli.backup.handler.ShellHandler.Companion.utilBoxQ
import com.machiav3lli.backup.handler.ShellHandler.ShellCommandFailedException
import com.machiav3lli.backup.items.Package
import com.machiav3lli.backup.items.RootFile
import com.machiav3lli.backup.items.StorageFile
import com.machiav3lli.backup.tasks.AppActionWork
import com.machiav3lli.backup.utils.CryptoSetupException
import com.machiav3lli.backup.utils.suUnpackTo
import org.apache.commons.compress.archivers.tar.TarArchiveInputStream
import org.apache.commons.io.FileUtils
import timber.log.Timber
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException

class RestoreSpecialAction(context: Context, work: AppActionWork?, shell: ShellHandler) :
    RestoreAppAction(context, work, shell) {

    @Throws(CryptoSetupException::class, RestoreFailedException::class)
    override fun restoreAllData(
        work: AppActionWork?,
        app: Package,
        backup: Backup,
        backupDir: StorageFile,
        backupMode: Int
    ) {
        work?.setOperation("dat")
        restoreData(app, backup, backupDir)
    }

    @Throws(RestoreFailedException::class, CryptoSetupException::class)
    override fun restoreData(
        app: Package,
        backup: Backup,
        backupDir: StorageFile
    ) {
        Timber.i("%s: Restore special data", app)
        val metaInfo = app.packageInfo as SpecialInfo
        val tempPath = RootFile(context.cacheDir, backup.packageName)
        val backupFilename = getBackupArchiveFilename(
            BACKUP_DIR_DATA,
            backup.isCompressed,
            backup.isEncrypted
        )
        val backupArchiveFile = backupDir.findFile(backupFilename)
            ?: throw RestoreFailedException("Backup archive at $backupFilename is missing")
        try {
            TarArchiveInputStream(
                openArchiveFile(backupArchiveFile, backup.isCompressed, backup.isEncrypted, backup.iv)
            ).use { archiveStream ->
                tempPath.mkdir()
                // Extract the contents to a temporary directory
                archiveStream.suUnpackTo(tempPath, isOldVersion(backup))

                // check if all expected files are there
                val filesInBackup = tempPath.listFiles()
                val expectedFiles = metaInfo.specialFiles
                    .map { pathname: String? -> RootFile(pathname ?: "") }
                    .toTypedArray()
                if (filesInBackup != null
                    && (filesInBackup.size != expectedFiles.size
                            || !areBasefilesSubsetOf(expectedFiles, filesInBackup))
                ) {
                    val errorMessage =
                        "$app: Backup is missing files. Found ${filesInBackup.map { it.absolutePath }}; needed: ${expectedFiles.map { it.absolutePath }}"
                    Timber.e(errorMessage)
                    throw RestoreFailedException(errorMessage, null)
                }
                val commands = mutableListOf<String?>()
                for (restoreFile in expectedFiles) {
                    val (uid, gid, con) = try {
                        shell.suGetOwnerGroupContext(restoreFile.absolutePath)
                    } catch (e: Throwable) {
                        // fallback to permissions of parent directory
                        shell.suGetOwnerGroupContext(
                            restoreFile.parentFile?.absolutePath
                                ?: restoreFile.toPath().parent.toString()
                        )
                    }
                    commands.add(
                        "$utilBoxQ mv -f ${
                            quote(
                                File(
                                    tempPath,
                                    restoreFile.name
                                )
                            )
                        } ${quote(restoreFile)}"
                    )
                    commands.add(
                        "$utilBoxQ chown $uid:$gid ${quote(restoreFile)}"
                    )
                    commands.add(
                        if (con == "?") //TODO hg42: when does it happen? maybe if selinux not supported on storage?
                            null // "" ; restorecon -RF -v ${quote(restoreFile)}"  //TODO hg42 doesn't seem to work, probably because selinux unsupported in this case
                        else
                            "chcon -R -h -v '$con' ${quote(restoreFile)}"
                    )
                }
                val command = commands.filterNotNull().joinToString(" ; ")  // no dependency
                runAsRoot(command)
            }
            if (app.packageName == "special.smsmms.json") {
                for (filePath in metaInfo.specialFiles) {
                    RestoreSMSMMSJSONAction.restoreData(context, filePath)
                }
            }
            if (app.packageName == "special.calllogs.json") {
                for (filePath in metaInfo.specialFiles) {
                    RestoreCallLogsJSONAction.restoreData(context, filePath)
                }
            }
        } catch (e: RuntimeException) {
            throw RestoreFailedException("${e.message}", e)
        } catch (e: ShellCommandFailedException) {
            val error = extractErrorMessage(e.shellResult)
            Timber.e("$app: Restore $BACKUP_DIR_DATA failed. System might be inconsistent: $error")
            throw RestoreFailedException(error, e)
        } catch (e: FileNotFoundException) {
            throw RestoreFailedException("Could not find backup archive", e)
        } catch (e: IOException) {
            Timber.e("$app: Restore $BACKUP_DIR_DATA failed with IOException. System might be inconsistent: $e")
            throw RestoreFailedException("IOException", e)
        } finally {
            val backupDeleted = FileUtils.deleteQuietly(tempPath)   // if deleteQuietly is missing, org.apache.commons.io is wrong (shitty version from 2003 that looks newer)
            Timber.d("$app: Uncompressed $BACKUP_DIR_DATA was deleted: $backupDeleted")
        }
        if (app.packageName == "special.smsmms.json" || app.packageName == "special.calllogs.json") {
            for (filePath in metaInfo.specialFiles) {
                File(filePath).delete()
            }
        }
    }

    override fun restorePackage(backupDir: StorageFile, backup: Backup) {
        // stub
    }

    override fun restoreDeviceProtectedData(
        app: Package,
        backup: Backup,
        backupDir: StorageFile
    ) {
        // stub
    }

    override fun restoreExternalData(
        app: Package,
        backup: Backup,
        backupDir: StorageFile
    ) {
        // stub
    }

    override fun restoreObbData(
        app: Package,
        backup: Backup,
        backupDir: StorageFile
    ) {
        // stub
    }

    override fun refreshAppInfo(context: Context, app: Package) {
        // stub
    }

    companion object {
        private fun areBasefilesSubsetOf(
            set: Array<RootFile>,
            subsetList: Array<RootFile>
        ): Boolean {
            val baseCollection: Collection<String> = set.map { obj: File -> obj.name }.toHashSet()
            val subsetCollection: Collection<String> =
                subsetList.map { obj: File -> obj.name }.toHashSet()
            return baseCollection.containsAll(subsetCollection)
        }
    }
}